community/logging/src/main/java/org/neo4j/logging/async/AsyncLogEvent.java

```diff
@@ -0,0 +1,178 @@
+
+            @Override
+            public String toString()
+            {
+                return "Log.bulkLog( " + consumer + " )";
+            }
+        } );
+    }
+
+    static AsyncLogEvent bulkLogEvent( Logger logger, final Consumer<Logger> consumer )
+    {
+        requireNonNull( consumer, "Consumer<Logger>" );
+        return new AsyncLogEvent( logger, null, new BulkLogger()
+        {
+            @Override
+            void process( long timestamp, Object target )
+            {
+                consumer.accept( (Logger) target ); // TODO: include timestamp!
+            }
+
+            @Override
+            public String toString()
+            {
+                return "Logger.bulkLog( " + consumer + " )";
+            }
+        } );
+    }
+
+    @SuppressWarnings( "StringEquality" )
+    public void process()
+    {
+        if ( parameter == null )
+        {
+            ((Logger) target).log( "[AsyncLog @ " + timestamp() + "]  " + message );
+        }
+        else if ( parameter instanceof Throwable )
+        {
+            ((Logger) target).log( "[AsyncLog @ " + timestamp() + "]  " + message, (Throwable) parameter );
+        }
+        else if ( parameter instanceof Object[] )
+        {
+            ((Logger) target).log( "[AsyncLog @ " + timestamp() + "]  " + message, (Object[]) parameter );
+        }
+        else if ( parameter instanceof BulkLogger )
+        {
+            ((BulkLogger) parameter).process( timestamp, target );
+        }
+    }
+
+    private final long timestamp;
+    private final Object target;
+    private final String message;
+    private final Object parameter;
+
+    private AsyncLogEvent( Object target, String message, Object parameter )
+    {
+        this.target = target;
+        this.message = message;
+        this.parameter = parameter;
+        this.timestamp = System.currentTimeMillis();
+    }
+
+    @Override
+    public String toString()
+    {
+        if ( parameter == null )
+        {
+            return "log( @ " + timestamp() + ": \"" + message + "\" )";
+        }
+        if ( parameter instanceof Throwable )
+        {
+            return "log( @ " + timestamp() + ": \"" + message + "\", " + parameter + " )";
+        }
+        if ( parameter instanceof Object[] )
+        {
+            return "log( @ " + timestamp() + ": \"" + message + "\", " +
+                   Arrays.toString( (Object[]) parameter ) + " )";
+        }
+        if ( parameter instanceof BulkLogger )
+        {
+            return parameter.toString();
+        }
+        return super.toString();
+    }
+
+    private String timestamp()
+    {
+        return TIMESTAMP.format( timestamp );
+    }
+
+    private static abstract class BulkLogger
+    {
+        abstract void process( long timestamp, Object target );
+    }
+
+    private static final ThreadLocalFormat TIMESTAMP = new ThreadLocalFormat();
+
+    private static class ThreadLocalFormat extends ThreadLocal<DateFormat>
```

Comment:
I know, I wrote it. `org.neo4j.helpers` is in the `kernel` module, the `logging` module does not depend on the kernel module. It didn't seem like a thing worth rethinking the module structure for at this point.